google-chart-loader.js

```diff
@@ -0,0 +1,385 @@
+import '@polymer/polymer/polymer-legacy.js';
+import { Polymer } from '@polymer/polymer/lib/legacy/polymer-fn.js';
+
+/** @type {string} Most charts use this package. */
+var DEFACTO_CHART_PACKAGE = 'corechart';
+
+/** @enum {string} Namespaces that contain chart constructors. */
+var Namespace = {
+  CHARTS: 'charts',
+  VIS: 'visualization'
+};
+
+/**
+ * A collection of chart type details.
+ *
+ * @type {!Object<string, {ctor: string, pkg: (string|undefined)}>}
+ */
+var CHART_CONSTRUCTORS = {
+  'area': {
+    ctor: 'AreaChart',
+  },
+  'bar': {
+    ctor: 'BarChart',
+  },
+  'md-bar': {
+    ctor: 'Bar',
+    pkg: 'bar',
+  },
+  'bubble': {
+    ctor: 'BubbleChart',
+  },
+  'calendar': {
+    ctor: 'Calendar',
+    pkg: 'calendar',
+  },
+  'candlestick': {
+    ctor: 'CandlestickChart',
+  },
+  'column': {
+    ctor: 'ColumnChart',
+  },
+  'combo': {
+    ctor: 'ComboChart',
+  },
+  'gauge': {
+    ctor: 'Gauge',
+    pkg: 'gauge',
+  },
+  'geo': {
+    ctor: 'GeoChart',
+  },
+  'histogram': {
+    ctor: 'Histogram',
+  },
+  'line': {
+    ctor: 'LineChart',
+  },
+  'md-line': {
+    ctor: 'Line',
+    pkg: 'line',
+  },
+  'org': {
+    ctor: 'OrgChart',
+    pkg: 'orgchart',
+  },
+  'pie': {
+    ctor: 'PieChart',
+  },
+  'sankey': {
+    ctor: 'Sankey',
+    pkg: 'sankey',
+  },
+  'scatter': {
+    ctor: 'ScatterChart',
+  },
+  'md-scatter': {
+    ctor: 'Scatter',
+    pkg: 'scatter',
+  },
+  'stepped-area': {
+    ctor: 'SteppedAreaChart',
+  },
+  'table': {
+    ctor: 'Table',
+    pkg: 'table',
+  },
+  'timeline': {
+    ctor: 'Timeline',
+    pkg: 'timeline',
+  },
+  'treemap': {
+    ctor: 'TreeMap',
+    pkg: 'treemap',
+  },
+  'wordtree': {
+    ctor: 'WordTree',
+    namespace: Namespace.VIS,
+    pkg: 'wordtree',
+  }
+};
+
+/**
+ * The Google Charts constructor namespace inferred from the given chart type.
+ * @param {string} type the type of the chart
+ * @return {!Object} the namespace that contains the chart's constructor
+ */
+function namespaceForType(type) {
+  return google[type.indexOf('md-') === 0 ? Namespace.CHARTS : Namespace.VIS];
+}
+
+/** @type {!Object<string, boolean>} set-like object of gviz packages to load */
+var packagesToLoad = {};
+/** @type {!Object<string, !Promise>} promises for the various packages */
+var promises = {};
+/** @type {!Object<string, function(!Object)>} resolves for the package promises */
+var resolves = {};
+
+let resolveInitialize;
+let rejectInitialize;
+/** @type {!Promise<undefined>} */
+const initialize = new Promise((resolve, reject) => {
+  resolveInitialize = resolve;
+  rejectInitialize = reject;
+});
+
+/**
+ * Initializes Google Charts loader.
+ * 
+ * This function must be called only once. If it is called more than once, it throws an error.
+ */
+export function initializeOnce() {
+  if (!resolveInitialize) {
+    throw new Error('google-chart-loader already initialized');
```

Comment:
perhaps just return?

Reply:
I plan to add `options` parameter to this method to allow some one-time configuration. With that calling this method multiple times (with potentially different settings) is a mistake. I think it is better to throw an error than silently ignore the call.

Reply:
This is another reason I think that we shouldn't have the function. A single global can be used as the setup options.

Reply:
I discussed that approach with @rictic. The issue with using a global is that the users need to ensure that the code is loaded in the correct order, i.e. setting the global property happens before _any_ chart tries to load.

The explicit initialization function can be called at any time. That gives the user greater control. They can even defer the initialization to a later point.

Reply:
I think it's quite common and acceptable to set these options in the head of the root document. Another thing is that, in general, these options shouldn't be set. Having the initialize function adds a complication that need not exist.

Using the document's language and `current` version fits most use cases. If not, there would definitely be cases where the function is called more than once and subsequent invocations throw. Maybe they're resolvable, maybe not.

For the exceptions, I think globals are fine. As a user, I hate having to call initialize functions.